numhask/src/NumHask/Data/LogField.hs

```diff
@@ -0,0 +1,500 @@
+newtype LogField a = LogField a
+      deriving (Eq, Ord, Read, Data, Generic, Generic1, Functor, Foldable, Traversable)
+    -- TODO original has comment: , Ord -- Should we really perpetuate the Ord lie?
+    -- why is the comment there?
+
+----------------------------------------------------------------
+-- | Constructor which does semantic conversion from normal-domain
+-- to log-domain. Throws errors on negative and NaN inputs. If @p@
+-- is non-negative, then following equivalence holds:
+--
+-- > logField p == logToLogField (log p)
+logField :: (ExpField a) => a -> LogField a
+{-# INLINE [0] logField #-}
+-- TODO: should we use NOINLINE or [~0] to avoid the possibility of code bloat?
+logField = LogField . log
+
+
+-- TODO: figure out what to do here, removed guards
+-- | Constructor which assumes the argument is already in the
+-- log-domain.
+logToLogField :: a -> LogField a
+logToLogField = LogField
+
+
+-- | Semantically convert our log-domain value back into the
+-- normal-domain. Beware of overflow\/underflow. The following
+-- equivalence holds (without qualification):
+--
+-- > fromLogField == exp . logFromLogField
+--
+fromLogField :: ExpField a => LogField a -> a
+{-# INLINE [0] fromLogField #-}
+-- TODO: should we use NOINLINE or [~0] to avoid the possibility of code bloat?
+fromLogField (LogField x) = exp x
+
+
+-- | Return the log-domain value itself without conversion.
+logFromLogField :: LogField a -> a
+logFromLogField (LogField x) = x
+
+
+-- These are our module-specific versions of "log\/exp" and "exp\/log";
+-- They do the same things but also have a @LogField@ in between
+-- the logarithm and exponentiation. In order to ensure these rules
+-- fire, we have to delay the inlining on two of the four
+-- con-\/destructors.
+
+{-# RULES
+-- Out of log-domain and back in
+"log/fromLogField"       forall x. log (fromLogField x) = logFromLogField x
+"LogField/fromLogField"  forall x. LogField (fromLogField x) = x
+
+-- Into log-domain and back out
+"fromLogField/LogField"  forall x. fromLogField (LogField x) = x
+    #-}
+
+----------------------------------------------------------------
+-- To show it, we want to show the normal-domain value rather than
+-- the log-domain value. Also, if someone managed to break our
+-- invariants (e.g. by passing in a negative and noone's pulled on
+-- the thunk yet) then we want to crash before printing the
+-- constructor, rather than after.  N.B. This means the show will
+-- underflow\/overflow in the same places as normal doubles since
+-- we underflow at the @exp@. Perhaps this means we should show the
+-- log-domain value instead.
+
+instance (ExpField a, Show a) => Show (LogField a) where
+    showsPrec p (LogField x) =
+        let y = exp x in y `seq`
+        showParen (p > 9)
+            ( showString "LogField "
+            . showsPrec 11 y
+            )
+
+
+-- TODO: never worked with FFI, so i don't really know what's going on
```

Comment:
No idea either.  Let's ditch it all for now.  If I had to guess, there may be some tricky specialisation stuff we can do when working with doubles.

Reply:
Probably, but I don't think this is the concern of the module